extension/InspectionDoc/src/Infrastructure/Repository/InspectionDocRepository.php

```diff
@@ -0,0 +1,141 @@
+     */
+    public function __construct(InspectionDocDataProviderInterface $inspectionDocDataProvider)
+    {
+        $this->inspectionDocDataProvider = $inspectionDocDataProvider;
+    }
+
+    /**
+     * @param string $errorCode
+     *
+     * @return \InspectionDoc\Entity\InspectionDocInterface|null
+     */
+    public function findByErrorCode(string $errorCode): ?InspectionDocInterface
+    {
+        $indexedInspectionDocs = $this->getIndexedInspectionDocs();
+
+        $inspectionId = $this->findInspectionIdByErrorCode(array_keys($indexedInspectionDocs), $errorCode);
+
+        if ($inspectionId === null) {
+            return null;
+        }
+
+        $inspectionDoc = $indexedInspectionDocs[$inspectionId];
+
+        return $this->createInspectionDoc($inspectionId, $inspectionDoc);
+    }
+
+    /**
+     * @param array $inspectionsIds
+     * @param string $errorCode
+     *
+     * @return string|null
+     */
+    protected function findInspectionIdByErrorCode(array $inspectionsIds, string $errorCode): ?string
+    {
+        $inspectionIdCandidates = [];
+
+        foreach ($inspectionsIds as $inspectionsId) {
+            if (!$this->strStartsWith($errorCode, $inspectionsId)) {
+                continue;
+            }
+
+            $inspectionIdCandidates[] = $inspectionsId;
+        }
+
+        return $this->getMaxLengthMatch($inspectionIdCandidates);
+    }
+
+    /**
+     * @param string $haystack
+     * @param string $needle
+     *
+     * @return bool
+     */
+    protected function strStartsWith(string $haystack, string $needle): bool
+    {
+        return strncmp($haystack, $needle, strlen($needle)) === 0;
+    }
+
+    /**
+     * @param array<string> $inspectionsIds
+     *
+     * @return string|null
+     */
+    protected function getMaxLengthMatch(array $inspectionsIds): ?string
+    {
+        usort($inspectionsIds, static fn (string $a, string $b): int => -1 * (strlen($a) <=> strlen($b)));
+
+        return $inspectionsIds[0] ?? null;
+    }
+
+    /**
+     * @return array<string, array<mixed>>
+     */
+    protected function getIndexedInspectionDocs(): array
+    {
+        if ($this->indexedInspectionDocs !== null) {
+            return $this->indexedInspectionDocs;
+        }
+
+        $inspectionDocs = $this->inspectionDocDataProvider->getInspectionDocs();
+
+        $this->indexedInspectionDocs = [];
+
+        foreach ($inspectionDocs as $inspectionDoc) {
+            if (!isset($inspectionDoc['inspectionId'])) {
+                continue;
+            }
+
+            $this->indexedInspectionDocs[$inspectionDoc['inspectionId']] = $inspectionDoc;
+        }
+
+        return $this->indexedInspectionDocs;
+    }
+
+    /**
+     * @param string $inspectionId
+     * @param array $inspectionDocData
+     *
+     * @return \InspectionDoc\Entity\InspectionDocInterface
+     */
+    protected function createInspectionDoc(string $inspectionId, array $inspectionDocData): InspectionDocInterface
```

Comment:
As it has the potential to have a bug because of the state. Also, it breaks SRP principle violation, I'd suggest extracting it to the `InspectionDocCollectionCacheStorage` class and cover with unit tests.

Reply:
I'm not sure we should create a helper only for one case. I don't mind having some helper functions if we have multiple occasions for it.